### PR TITLE
Use a signed message via MetaMask to decrypt persisted commits

### PR DESCRIPTION
**Pull request by ptare**
The committed values should now show up in the "Commit" and "Reveal"
states.

Issue #532.

Signed-off-by: Prasad Tare <prasad.tare@gmail.com>

### CODE REVIEW

**Comment by ptare** [comment]
So very sorry this PR is such a mess :(

File: voter-dapp/src/ActiveRequests.js [hunk 1]
```diff
@@ -49,8 +51,87 @@ function ActiveRequests() {
...
+        [account]: { ...prev[account], [currentRoundId]: messageSigningProcessingToken }
+      }));
+      // TODO(ptare): Handle the user refusing to sign the message.
+      const { privateKey } = await deriveKeyPairFromSignatureMetamask(web3, getKeyGenMessage(currentRoundId), account);
+      setDecryptionKeys(prev => ({ ...prev, [account]: { ...prev[account], [currentRoundId]: privateKey } }));
+    }
+
+    getDecryptionKey();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [account, currentRoundId, decryptionKeys]);
```
Comment (mrice32) [inline voter-dapp/src/ActiveRequests.js]: As discussed IRL, is `decryptionKeys` required here?

Also, if this memoization works correctly (no spurious calls), could we completely remove the `messageSigningProcessingToken` stage of this?

File: common/Crypto.js [hunk 2]
```diff
@@ -27,21 +27,51 @@ async function decryptMessage(privKey, encryptedMessage) {
...
+async function getMessageSignatureMetamask(web3, messageToSign, signingAccount) {
+  return await web3.eth.personal.sign(messageToSign, signingAccount);
+}
+
+async function getMessageSignatureTruffle(web3, messageToSign, signingAccount) {
+  const signature = await web3.eth.sign(messageToSign, signingAccount);
+  // It's probably a good idea to add a comment here explaining why we have to do this manipulation, but I don't
+  // actually understand why.
+  const rAndS = signature.substring(0, 128 + 2);
+  const v = signature.substring(128 + 2, 132);
```
[the remark this answers is not in the document]
Comment (ptare) [reply common/Crypto.js]: Heh I've written what I *think* is going on, might be wrong though.

File: voter-dapp/src/ActiveRequests.js [hunk 3]
```diff
@@ -49,8 +51,87 @@ function ActiveRequests() {
     // Each of the subfetches has to complete. In drizzle, `undefined` means incomplete, while `null` means complete
     // but the fetched value was null, e.g., no `comittedValue` existed.
     voteStatuses.every(voteStatus => voteStatus.committedValue !== undefined && voteStatus.hasRevealed !== undefined);
+  // Future hook calls that depend on `voteStatuses` should use `voteStatusesStringified` in their dependencies array
+  // because `voteStatuses` will never compare equal after a re-render, even if no values in it actually changed.
+  // Also note that `JSON.stringify` doesn't distinguish `undefined` and `null`, but in Drizzle, those mean different
```
[the remark this answers is not in the document]
Comment (ptare) [reply voter-dapp/src/ActiveRequests.js]: Done.

File: voter-dapp/src/ActiveRequests.js (hunk 1 above)
Comment (ptare) [reply voter-dapp/src/ActiveRequests.js]: Done.

File: common/Crypto.js [hunk 4]
```diff
@@ -27,21 +27,51 @@ async function decryptMessage(privKey, encryptedMessage) {
...
 // Note: this is not meant to be used to generate private keys that hold ETH or any other high value assets. This is
 // meant to create a node/metamask friendly way of generating a temporary encryption/decryption key for sending private
 // messages.
+
+// Derive a private key, that works *only* on Metamask.
+async function deriveKeyPairFromSignatureMetamask(web3, messageToSign, signingAccount) {
+  return deriveKeyPair(web3, await getMessageSignatureMetamask(web3, messageToSign, signingAccount));
+}
+
+// Derive a private key, that works *only* with Truffle.
```
[the remark this answers is not in the document]
Comment (ptare) [reply common/Crypto.js]: I have some pending changes to Voting.js (which would need to be adjusted), do you mind if I do this rename in a separate PR?

[the remark this answers is not in the document]
Comment (mrice32) [reply common/Crypto.js]: SGTM

File: common/Crypto.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrice32) [reply common/Crypto.js]: This looks great! Thanks!

File: voter-dapp/src/ActiveRequests.js [hunk 5]
```diff
@@ -49,8 +51,78 @@ function ActiveRequests() {
...
-  if (!initialFetchComplete || !subsequentFetchComplete || !revealEvents) {
+  // The decryption key is a function of the account and `currentRoundId`, so we need the user to re-sign a message
+  // any time one of those changes. We also don't want to show multiple, identical notifications when this effect gets
+  // cleaned up on unrelated re-renders. The best solution I've come up with so far is to store the keys in a nested map
+  // (i.e., Object) from account => roundId => key. When a signing request is sent out (i.e., the user is shown a
+  // Metamask popup), we inject a special `messageSigningProcessingToken` at `decryptionKeys[account][currentRoundId]`
+  // to prevent resending that request. Note that in this approach, we don't disregard the result of this hook on
+  // component unmount.
+  const [decryptionKeys, setDecryptionKeys] = useState({});
+  useEffect(() => {
```
Comment (rawfalafel) [inline voter-dapp/src/ActiveRequests.js]: Can this function itself be `async`? Started taking a look at this: https://www.robinwieruch.de/react-hooks-fetch-data/

Comment (rawfalafel) [reply voter-dapp/src/ActiveRequests.js]: lol, nvm, I kept on reading the rest of this.

**Pull request merged**